lxd/storage/drivers/driver_btrfs_volumes.go

```diff
@@ -808,22 +808,45 @@ func (d *btrfs) RefreshVolume(vol VolumeCopy, srcVol VolumeCopy, refreshSnapshot
 		// Configure the pipes.
 		receiver.Stdin, _ = sender.StdoutPipe()
 		receiver.Stdout = os.Stdout
-		receiver.Stderr = os.Stderr
+
+		var recvStderr bytes.Buffer
+		receiver.Stderr = &recvStderr
 
 		// Run the transfer.
-		err = receiver.Start()
+		err := receiver.Start()
 		if err != nil {
-			return fmt.Errorf("Failed to receive stream: %w", err)
+			return fmt.Errorf("Failed starting BTRFS receive: %w", err)
 		}
 
-		err = sender.Run()
+		var sendStderr bytes.Buffer
+		sender.Stderr = &sendStderr
+		err = sender.Start()
 		if err != nil {
-			return fmt.Errorf("Failed to send stream %q: %w", sender.String(), err)
+			_ = receiver.Process.Kill()
+			return fmt.Errorf("Failed starting BTRFS send: %w", err)
 		}
 
+		senderErr := make(chan error)
+		go func() {
+			err := sender.Wait()
+			if err != nil {
+				_ = receiver.Process.Kill()
+				senderErr <- fmt.Errorf("Failed BTRFS send: %w (%s)", err, strings.TrimSpace(sendStderr.String()))
+				return
+			}
+
+			senderErr <- nil
+		}()
+
 		err = receiver.Wait()
 		if err != nil {
-			return fmt.Errorf("Failed to wait for receiver: %w", err)
+			_ = receiver.Process.Kill()
```

Comment:
I understand above if the sender fails, the receiver also gets killed. But here if the receiver fails, don't we want to kill the sender instead?

Reply:
yeah that looks like a bug in the zfs driver too im already working on fixing :)
Also the ZFS driver's RefreshVolume function looks to have the same issue wrt to hanging as the BTRFS driver so making them all consistent.